yarn-project/telemetry-client/src/telemetry.ts

```diff
@@ -62,8 +67,98 @@ export interface TelemetryClient {
    */
   getMeter(name: string): Meter;
 
+  /**
+   * Creates a new tracer
+   * @param name - The name of the tracer.
+   */
+  getTracer(name: string): Tracer;
+
   /**
    * Stops the telemetry client.
    */
   stop(): Promise<void>;
 }
+
+/** Objects that adhere to this interface can use @trackSpan */
+export interface Traceable {
+  tracer: Tracer;
+}
+
+type SpanDecorator<T extends Traceable, F extends (...args: any[]) => any> = (
+  originalMethod: F,
+  context: ClassMethodDecoratorContext<T>,
+) => F;
+
+/**
+ * Starts a new span whenever the decorated method is called.
+ * @param spanName - The name of the span to create. Can be a string or a function that returns a string.
+ * @param attributes - Initial attributes to set on the span. If a function is provided, it will be called with the arguments of the method.
+ * @param extraAttributes - Extra attributes to set on the span after the method is called. Will be called with the return value of the method. Note: if the function throws then this will not be called.
+ * @returns A decorator that wraps the method in a span.
+ */
+export function trackSpan<T extends Traceable, F extends (...args: any[]) => any>(
+  spanName: string | ((this: T, ...args: Parameters<F>) => string),
+  attributes?: Attributes | ((this: T, ...args: Parameters<F>) => Attributes),
+  extraAttributes?: (this: T, returnValue: Awaited<ReturnType<F>>) => Attributes,
+): SpanDecorator<T, F> {
+  return (originalMethod: F, _context: ClassMethodDecoratorContext<T>) => {
+    return function replacementMethod(this: T, ...args: Parameters<F>): Promise<Awaited<ReturnType<F>>> {
+      const name = typeof spanName === 'function' ? spanName.call(this, ...args) : spanName;
+      const currentAttrs = typeof attributes === 'function' ? attributes.call(this, ...args) : attributes;
+      return this.tracer.startActiveSpan(name, async (span: Span) => {
+        for (const [key, value] of Object.entries(currentAttrs ?? {})) {
+          span.setAttribute(key, value);
+        }
+
+        try {
+          const res = await originalMethod.call(this, ...args);
+          const extraAttrs = extraAttributes?.call(this, res);
+          for (const [key, value] of Object.entries(extraAttrs ?? {})) {
+            span.setAttribute(key, value);
+          }
+
+          return res;
+        } catch (err) {
+          span.setStatus({
+            code: SpanStatusCode.ERROR,
+            message: String(err),
+          });
+          throw err;
+        } finally {
+          span.end();
+        }
+      });
+    } as F;
+  };
+}
+
+/**
+ * Runs an event callback in a span. The span is started immediately and completes once the callback finishes running.
```

Comment:
Shouldn't the span be started when the callback function is invoked by the callee, rather than immediately?

Reply:
Potentially, but this also accounts for the delay in scheduling the function to be called by the orchestrator. I'll give it a try.

Reply:
I've added two events to this span: immediately before the callback is called and right after it finishes. One interesting instance where there was a 300ms delay between scheduling the proving request and actually running (span start time 26ms, callback start event 322ms)

![image](https://github.com/AztecProtocol/aztec-packages/assets/3816165/3ee8d8a2-fe3a-4c97-a7f7-98bd81dbdd76)